Move app/device to session interface

fractalwrench opened the pull request:
## Goal

Moves app/device to the `Session` interface and removes batch delivery of sessions. This makes the notifier conform to the notifier spec.

## Changeset

- Removed `SessionPayload` from public interface
- Added `app/device` onto `Session`
- Updated associated code to set app/device on `Session`, and to deliver `Session`
- Updated `Event` serializer to handle null `startedAt` value
- `SessionTracker` now only delivers one cached file at a time
- Simplified `Session` constructor and made it possible to initialise from a `File`
- `Session` now serializes the entire JSON payload (from a POJO or a file)

## Tests

Updated existing tests to match changeset, and added new coverage in `SessionTest` for new API fields.

fractalwrench commented on bugsnag-android-core/src/main/java/com/bugsnag/android/Session.java:
```diff
@@ -2,21 +2,32 @@
 
 import androidx.annotation.NonNull;
 
+import org.jetbrains.annotations.Nullable;
```
Nope, I've updated this to use the right import, rather than the auto-suggested one.

fractalwrench commented on bugsnag-android-core/src/main/java/com/bugsnag/android/Session.java:
```diff
@@ -26,39 +37,58 @@ static Session copySession(Session session) {
+    }
+
+    public void setStartedAt(@Nullable Date startedAt) {
+        this.startedAt = startedAt;
```
I believed these were nullable on the spec but can see they are not - updated to `NonNull`.

fractalwrench commented on bugsnag-android-core/src/main/java/com/bugsnag/android/Session.java:
```diff
@@ -26,39 +36,58 @@ static Session copySession(Session session) {
+
+    void setDevice(Device device) {
+        this.device = device;
+    }
```
`app` and `device` are readonly fields on the spec so setters are not required - I will add missing getters.

fractalwrench commented on bugsnag-android-core/src/main/java/com/bugsnag/android/SessionTracker.java:
```diff
@@ -233,16 +232,10 @@ Session incrementHandledAndCopy() {
-                            new SessionPayload(null, storedFiles,
-                                    client.appDataCollector.generateApp(),
-                                    client.deviceDataCollector.generateDevice());
+                List<File> storedFiles = sessionStore.findStoredFiles();
```
Good point. Currently this implementation would just attempt to deliver any cached sessions as-is, and they would be dropped by the pipeline. Depending on the impact of this we could choose to send these legacy payloads in the same way as before by distinguishing the new payloads in some way. For instance, we could:

- store v5 sessions in a different location
- encode version information into the payload (e.g. add a v5 suffix)

tomlongridge replied:
As discussed IRL, let's go for the latter and encode a file version into the file name so that this pattern can be re-used in future if required.

fractalwrench commented:
I've updated the PR to address the feedback. Changes include:

- Added setters for `session.app` and `session.device`
- Encoded "v2" in the session payload for new files. This invokes a different serialization path if the `Session` class detects a legacy cached file is present
- Added instrumentation tests to verify that V1 and V2 payloads serialize correctly
- Extracted `flushStoredSession()` in the `SessionTracker`, fixing a bug where _all_ the queued reports would be deleted after one was delivered

tomlongridge commented on bugsnag-android-core/src/main/java/com/bugsnag/android/Session.java:
```diff
@@ -89,15 +128,56 @@ void setAutoCaptured(boolean autoCaptured) {
+        writer.value(file);
+    }
+
+    private void serializeV1Payload(@NonNull JsonStream writer) throws IOException { // TODO test me
```
See TODO comment :)

fractalwrench replied:
This is addressed by existing test coverage in `SessionV1PayloadTest` - I just forgot to remove this comment.

fractalwrench commented on bugsnag-android-core/src/test/java/com/bugsnag/android/SessionTest.kt:
```diff
@@ -33,11 +37,41 @@ class SessionTest {
         validateSessionCopied(copy)
     }
 
+    @Test
```
Added a test for the getter/setter

fractalwrench commented on bugsnag-android-core/src/test/java/com/bugsnag/android/SessionSerializationTest.kt:
```diff
@@ -13,10 +15,15 @@ internal class SessionSerializationTest {
-            "session",
-            Session("123", Date(0), User("123", "j@example.com", "Jay"), 1, 0)
-        )
+        fun testCases(): Collection<Pair<Any, String>> {
```
Using null/placeholder values for complex nested objects is something we agreed to do when we originally did this work. The reasoning behind this was that the test cases will become more complex for minimal benefit, and that we should just verify that _something_ is serialized for the user key. We have existing coverage to check how a user is serialized in `UserSerializationTest` which I think is ok to rely on, unless you feel strongly about this.

tomlongridge commented on bugsnag-android-core/src/main/java/com/bugsnag/android/SessionTracker.java:
```diff
@@ -233,42 +233,46 @@ Session incrementHandledAndCopy() {
+    void flushStoredSession(File storedFile) {
+        Session payload = new Session(storedFile);
+
+        if (payload.isV2Payload()) { // collect data here
```
I'm a little concerned that this wasn't picked up in any of the unit/manual tests. I guess that's an existing hole in the flushStoredSession functionality in SessionTrackerTests.
